clippy_lints/src/methods/unused_enumerate_index.rs

```diff
@@ -0,0 +1,113 @@
+/// * `recv`: The receiver of the call
+/// * `closure_arg`: The argument to the map function call containing the closure/function to apply
+pub(super) fn check(cx: &LateContext<'_>, call_expr: &Expr<'_>, recv: &Expr<'_>, closure_arg: &Expr<'_>) {
+    let recv_ty = cx.typeck_results().expr_ty(recv);
+    if let Some(recv_ty_defid) = recv_ty.ty_adt_def().map(AdtDef::did)
+        // If we call a method on a `std::iter::Enumerate` instance
+        && match_def_path(cx, recv_ty_defid, &CORE_ITER_ENUMERATE_STRUCT)
+        // If we are calling a method of the `Iterator` trait
+        && is_trait_method(cx, call_expr, sym::Iterator)
+        // And the map argument is a closure
+        && let ExprKind::Closure(closure) = closure_arg.kind
+        && let closure_body = cx.tcx.hir().body(closure.body)
+        // And that closure has one argument ...
+        && let [closure_param] = closure_body.params
+        // .. which is a tuple of 2 elements
+        && let PatKind::Tuple([index, elem], ..) = closure_param.pat.kind
+        // And that the first element (the index) is either `_` or unused in the body
+        && pat_is_wild(cx, &index.kind, closure_body)
+    {
+        // First, try to trigger the lint with our receiver. This will work when chaining methods:
+        // ```
+        // iter.enumerate().map(|(_, x)| x)
+        // ^^^^^^^^^^^^^^^^ `recv`, a call to `std::iter::Iterator::enumerate`
+        // ````
+        if !trigger_lint(cx, recv, closure_param.span, elem.span) {
+            // Otherwise, it may be that we have a binding. Find its initializer:
+            // ```
+            // let binding = iter.enumerate();
+            //               ^^^^^^^^^^^^^^^^ `recv_init_expr`
+            // binding.map(|(_, x)| x)
+            // ^^^^^^^ `recv`, not a call to `std::iter::Iterator::enumerate`
+            let recv_init_expr = expr_or_init(cx, recv);
+            trigger_lint(cx, recv_init_expr, closure_param.span, elem.span);
+
+            // Otherwise, the `Enumerate` probably comes from something that we cannot control.
+            // This would for instance happen with:
+            // ```
+            // external_lib::some_function_returning_enumerate().map(|(_, x)| x)
+            // ```
+        }
+    }
+}
+
+/// Trigger the lint with the suggestion.
+fn trigger_lint(cx: &LateContext<'_>, recv: &Expr<'_>, closure_param_span: Span, elem_span: Span) -> bool {
+    // Check if our `recv` is a call to `std::iter::Iterator::enumerate`.
+    if let ExprKind::MethodCall(_, enumerate_recv, _, enumerate_span) = recv.kind
+        && let Some(enumerate_defid) = cx.typeck_results().type_dependent_def_id(recv.hir_id)
+        && match_def_path(cx, enumerate_defid, &CORE_ITER_ENUMERATE_METHOD)
+    {
+        // If it is, we can suggest removing the tuple from the closure and the preceding call to
+        // `enumerate`, whose span we can get from the `MethodCall`.
+        span_lint_and_then(
```

Comment:
Indeed! A nice corner case.

I would however assume that I need the `hir_id` of the return of `expr_or_init(cx, recv)` rather than that of `recv`?

Writing the test made me notice that I do not handle the following case:
```rs
[].iter().enumerate().map(|_| {})
```
I'm gonna fix this and add it as a test before I push.

Reply:
Hmm. No, I am not gonna fix this in this PR, but rather with the next one where I also handle `.map(|x| x.1)`.